### PR TITLE
Initial commit for config files annotations

### PR DESCRIPTION
**Pull request by Josh-01**
Added change to include following to annotations:

1. Manifest files path
2. Helm chart path
3. For each image:
   - Dockerfile path
   - CR server name

Issues: https://github.com/github/GitHub-EcoSystem/issues/690

### CODE REVIEW

File: src/utilities/utility.ts [hunk 1]
```diff
@@ -127,6 +129,86 @@ export function annotateChildPods(kubectl: Kubectl, resourceType: string, resour
...
+                    throw new Error(`docker images pull failed with: ${res.stderr.match(/(.*)\s*$/)![0]}`);
+                }
+            });
+
+            await exec.exec('docker inspect --type=image', args, true).then(res => {
+                if (res.stderr != '' && !res.success) {
+                    throw new Error(`docker inspect call failed with: ${res.stderr.match(/(.*)\s*$/)![0]}`);
+                }
+
+                if(res.stdout!= null && res.stdout != ''){
```
Comment (koushdey) [inline src/utilities/utility.ts:183]: Modify the checks to `if (!res.stdout)`. Check [this](https://stackoverflow.com/questions/5515310/is-there-a-standard-function-to-check-for-null-undefined-or-blank-variables-in?rq=1). Check for such changes at other places as well.

File: __tests__/run.test.ts [hunk 2]
```diff
@@ -321,7 +323,8 @@ test("deployment - deploy() - Annotate & label resources", async () => {
 
 test("deployment - deploy() - Annotate & label resources for a new workflow", async () => {
     process.env.GITHUB_WORKFLOW = '.github/workflows/NewWorkflow.yml';
-    let annotationKeyValStr = getWorkflowAnnotationKeyLabel(process.env.GITHUB_WORKFLOW) + '=' + getWorkflowAnnotationsJson('NA', '.github/workflows/NewWorkflow.yml');
+    let filepathConfigs = { manifestFilePaths :['manifests/deployment.yaml'], helmChartFilePaths : '', buildConfigs :{}} ;
+    let annotationKeyValStr = getWorkflowAnnotationKeyLabel(process.env.GITHUB_WORKFLOW) + '=' + getWorkflowAnnotationsJson('NA', '.github/workflows/NewWorkflow.yml', filepathConfigs);
     const KubernetesManifestUtilityMock = mocked(KubernetesManifestUtility, true);
```
Comment (koushdey) [inline __tests__/run.test.ts:328]: Add a new test case for your changes around dockerfile path and helm chart file paths.

File: src/utilities/utility.ts (hunk 1 above)
Comment (Josh-01) [reply src/utilities/utility.ts:183]: Done.

File: __tests__/run.test.ts (hunk 2 above)
Comment (Josh-01) [reply __tests__/run.test.ts:328]: Will add it in next iteration, after the paths json structures are final

File: src/utilities/utility.ts [hunk 3]
```diff
@@ -127,6 +129,86 @@ export function annotateChildPods(kubectl: Kubectl, resourceType: string, resour
...
+    let imageNames = core.getInput('images').split('\n');
+
+    for(const image of imageNames){
+        let args: string[] = [image];
+        let resultObj: any;
+        let buildConfigMap : any = {};
+        let containerRegistryName = image.toString().split('/')[0];
+
+        try{
+
```
[the remark this answers is not in the document]
Comment (Josh-01) [reply src/utilities/utility.ts:159]: Done

File: src/utilities/utility.ts [hunk 4]
```diff
@@ -127,6 +129,86 @@ export function annotateChildPods(kubectl: Kubectl, resourceType: string, resour
...
+    let inputManifestFiles = inputParams.manifests || [];
+    filePathsConfig[MANIFEST_PATHS_KEY] = inputManifestFiles;
+
+    let helmChartPath = process.env.HELM_CHART_PATH || '';
+    filePathsConfig[HELM_CHART_KEY] = helmChartPath;
+
+    //Fetch labels from each image
+    let imageToBuildConfigMap: any = {};
+    let imageNames = core.getInput('images').split('\n');
```
[the remark this answers is not in the document]
Comment (Josh-01) [reply src/utilities/utility.ts:150]: Yes, is there a particular case that I should be testing ?

File: src/utilities/utility.ts [hunk 5]
```diff
@@ -127,6 +129,86 @@ export function annotateChildPods(kubectl: Kubectl, resourceType: string, resour
...
+    const MANIFEST_PATHS_KEY = 'manifestFilePaths';
+    const HELM_CHART_KEY = 'helmChartFilePaths';
+    const DOCKERFILE_PATH_LABEL_KEY = 'dockerfile-path';
+    const DOCKERFILE_PATH_KEY = 'dockerfilePath';
+    const CONTAINER_REG_KEY = 'containerRegistryServer';
+
+    let inputManifestFiles = inputParams.manifests || [];
+    filePathsConfig[MANIFEST_PATHS_KEY] = inputManifestFiles;
+
+    let helmChartPath = process.env.HELM_CHART_PATH || '';
```
[the remark this answers is not in the document]
Comment (Josh-01) [reply src/utilities/utility.ts:145]: Done.

File: src/utilities/utility.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Josh-01) [reply src/utilities/utility.ts:150]: Done.

File: __tests__/run.test.ts [hunk 6]
```diff
@@ -38,7 +38,7 @@ const getAllPodsMock = {
...
+    'stdout': '{"apiVersion": "v1","kind": "Namespace","metadata": {"annotations": {"githubWorkflow_c11401b9d232942bac19cbc5bc32b42d": "{\'run\': \'202489005\',\'repository\': \'testUser/hello-kubernetes\',\'workflow\': \'workflow1\',\'jobName\': \'build-and-deploy\',\'createdBy\': \'testUser\',\'runUri\': \'https://github.com/testUser/hello-kubernetes/actions/runs/202489005\',\'commit\': \'currentCommit\',\'lastSuccessRunCommit\': \'lastCommit\',\'branch\': \'refs/heads/branch-rename\',\'deployTimestamp\': \'1597062957973\',\'filePathConfigs\': \'{}\',\'provider\': \'GitHub\'}","githubWorkflow_21fd7a597282ca5adc05ba99018b3706": "{\'run\': \'202504411\',\'repository\': \'testUser/hello-kubernetes\',\'workflow\': \'workflowMaster\',\'jobName\': \'build-and-deploy\',\'createdBy\': \'testUser\',\'runUri\': \'https://github.com/testUser/hello-kubernetes/actions/runs/202504411\',\'commit\': \'currentCommit1\',\'lastSuccessRunCommit\': \'NA\',\'branch\': \'refs/heads/master\',\'deployTimestamp\': \'1597063919873\',\'filePathConfigs\': \'{}\',\'provider\': \'GitHub\'}"}},"spec": {"finalizers": ["kubernetes"]},"status": {"phase": "Active"}}'
```
Comment (koushdey) [inline __tests__/run.test.ts:41]: Change the mock response as per the new format updated with this PR.

File: src/utilities/utility.ts [hunk 7]
```diff
@@ -4,6 +4,9 @@ import { IExecSyncResult } from './tool-runner';
 import { Kubectl } from '../kubectl-object-model';
 import { GitHubClient } from '../githubClient';
 import { StatusCodes } from "./httpClient";
+import * as exec from "./exec";
+import * as inputParams from "../input-parameters";
+import { Z_FILTERED } from 'zlib';
```
Comment (koushdey) [inline src/utilities/utility.ts:9]: Remove this import as discussed

File: src/utilities/utility.ts [hunk 8]
```diff
@@ -127,6 +130,81 @@ export function annotateChildPods(kubectl: Kubectl, resourceType: string, resour
...
+    const DOCKERFILE_PATH_LABEL_KEY = 'dockerfile-path';
+
+    let inputManifestFiles = inputParams.manifests || [];
+    filePathsConfig[MANIFEST_PATHS_KEY] = inputManifestFiles;
+
+    let helmChartPaths = (process.env.HELM_CHART_PATHS && process.env.HELM_CHART_PATHS.split('\n').filter(path => path != "")) || [];
+    filePathsConfig[HELM_CHART_KEY] = helmChartPaths;
+
+    //Fetch labels from each image
+
```
Comment (koushdey) [inline src/utilities/utility.ts:148]: NIT: When editing test cases remove some of the new lines here.

**Review by koushdey — APPROVED** [review]
Make sure the new annotation object is not breaking while rendering.

File: __tests__/run.test.ts (hunk 6 above)
Comment (Josh-01) [reply __tests__/run.test.ts]: Done

File: src/utilities/utility.ts (hunk 8 above)
Comment (Josh-01) [reply src/utilities/utility.ts]: Done

**Comment by ds-ms** [comment]
Cherry pick these changes to `master` and `releases/v1.*` > v1.2 as well  😄

**Comment by Josh-01** [comment]
Discussed it with @ammohant  - we will be adding this to v1.2 and v1.3. 
From v1.4 onwards we will have the new deployment resource changes.

**Pull request merged**